File: user_interface/src/contracts/capabilities.ts
/** Features and resources confirmed by the running backend. */
export interface Capabilities {
  application_version: string;
  available_devices: string[];
  available_models: string[];
  available_profiles: string[];
  encoding_available: boolean;
  decoding_available: boolean;
  training_available: boolean;
  experimental_models_only: true;
  maximum_payload_bytes: number;
  minimum_image_side: number;
  maximum_image_side: number;
  maximum_upload_bytes: number;
  maximum_dataset_upload_bytes: number;
  dataset_upload_chunk_bytes: number;
  hugging_face_token_configured: boolean;
  dataset_source_kinds: DatasetSourceKind[];
}

/** Dataset sources the backend accepts in the Train tab. */
export type DatasetSourceKind =
  | 'server_folder'
  | 'upload'
  | 'hugging_face'
  | 'https_archive';

/** Backend readiness, separate from model availability. */
export interface HealthStatus {
  status: 'ready';
  application_version: string;
}

File: schemas/capabilities.py
"""Available application features for this local foundation release."""

from typing import Annotated, Literal

from pydantic import Field, field_validator

from schemas.base import StrictRecord


class HealthStatus(StrictRecord):
    """Report successful application startup."""

    status: Literal["ready"] = "ready"
    application_version: Literal["0.1.0"] = "0.1.0"


class Capabilities(StrictRecord):
    """Separate experimental training from qualified inference support."""

    application_version: Literal["0.1.0"] = "0.1.0"
    available_devices: list[Literal["cpu"]] = Field(
        default=["cpu"], min_length=1, max_length=1
    )
    available_models: Annotated[list[str], Field(max_length=0)] = Field(
        default_factory=list
    )
    available_profiles: Annotated[list[str], Field(max_length=0)] = Field(
        default_factory=list
    )
    encoding_available: Literal[False] = False
    decoding_available: Literal[False] = False
    training_available: bool = True
    maximum_payload_bytes: Annotated[int, Field(ge=0, le=0)] = 0

    @field_validator(
        "encoding_available", "decoding_available", "training_available", mode="before"
    )
    @classmethod
    def validate_availability_type(cls, value: object) -> object:
        """Reject numeric values that compare equal to a boolean literal."""
        if type(value) is not bool:
            raise ValueError("Feature availability must be a boolean.")
        return value


class ModelList(StrictRecord):
    """Return an empty installed-model list until model support is available."""

    items: Annotated[list[str], Field(max_length=0)] = Field(default_factory=list)
